crates/emath/src/ts_transform.rs

```diff
@@ -0,0 +1,91 @@
+use crate::{Pos2, Rect, Vec2};
+
+/// Linearly transforms positions via a translation, then a scaling.
+///
+/// [`TSTransform`] first translates points, then scales them with the scaling origin
+/// at `0, 0` (the top left corner)
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[cfg_attr(feature = "bytemuck", derive(bytemuck::Pod, bytemuck::Zeroable))]
+pub struct TSTransform {
+    /// Translation amount.
+    pub translation: Vec2,
+
+    /// Scaling amount after translation, scaled around (0, 0).
+    pub scaling: f32,
+}
+
+impl Eq for TSTransform {}
+
+impl Default for TSTransform {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            translation: Vec2::ZERO,
+            scaling: 1.0,
+        }
+    }
+}
+
+impl TSTransform {
+    /// The translation is applied first, then scaling around 0, 0.
+    pub fn new(translation: Vec2, scaling: f32) -> Self {
+        Self {
+            translation,
+            scaling,
+        }
+    }
+
+    pub fn from_translation(translation: Vec2) -> Self {
+        Self::new(translation, 1.0)
+    }
+
+    pub fn from_scaling(scaling: f32) -> Self {
+        Self::new(Vec2::ZERO, scaling)
+    }
+
+    /// Reverses the transformation from screen space to layer space.
+    pub fn invert_pos(&self, pos: Pos2) -> Pos2 {
+        // First, reverse scale around 0, 0, then reverse transform.
+        let pos = pos / self.scaling;
+        pos - self.translation
+    }
+
+    /// Reverses the transformation from screen space to layer space.
+    pub fn invert_rect(&self, rect: Rect) -> Rect {
+        Rect::from_min_max(self.invert_pos(rect.min), self.invert_pos(rect.max))
+    }
```

Comment:
I think a better API is to have an `.inverse()` function, so that `t.invert_pos(x) == t.inverse().mul_pos(x)`

Reply:
I think this would require having the translation applied in the inverse order. If we generalize the transform internally to be a 3x3 matrix (to account for the affine transformation), then it would be possible to invert.